libp2p/autotls/acme.nim

```diff
@@ -0,0 +1,317 @@
+    let directoryResponse =
+      await HttpClientRequestRef.get(session, acmeServerURL & "/directory").get().send()
+    let directory = bytesToString(await directoryResponse.getBodyBytes()).parseJson()
+
+    let acc = new(ACMEAccount)
+    acc.status = status
+    acc.contact = contact
+    acc.kid = kid
+    acc.key = key
+    acc.session = session
+    acc.directory = directory
+    acc.acmeServerURL = acmeServerURL
+    return acc
+  except HttpError as exc:
+    raise newException(ACMEError, "Failed to connect to ACME server", exc)
+  except ValueError as exc:
+    raise newException(ACMEError, "Unable to parse JSON", exc)
+  except OSError as exc:
+    raise newException(ACMEError, "Unable to parse JSON", exc)
+  except IOError as exc:
+    raise newException(ACMEError, "Unable to parse JSON", exc)
+  except Exception as exc:
+    raise newException(
+      ACMEError, "Unexpected error occurred while getting ACME server directory: ", exc
+    )
+
+proc newNonce(
+    self: ref ACMEAccount
+): Future[string] {.async: (raises: [ACMEError, CancelledError]).} =
+  try:
+    let nonceURL = self.directory.getJSONField("newNonce").getStr
+    let resp = await HttpClientRequestRef.get(self.session, nonceURL).get().send()
+    return resp.headers.getString("Replay-Nonce")
+  except HttpError:
+    raise newException(ACMEError, "Failed to request new nonce from ACME server")
+
+# TODO: save n and e in account so we don't have to recalculate every time
+proc acmeHeader(
+    self: ref ACMEAccount, url: string, needsJwk: bool, kid: string = ""
+): Future[JsonNode] {.async: (raises: [ACMEError, CancelledError]).} =
+  # TODO: check if scheme is RSA
+  let pubkey = self.key.pubkey.rsakey
+  let nArray = @(getArray(pubkey.buffer, pubkey.key.n, pubkey.key.nlen))
+  let eArray = @(getArray(pubkey.buffer, pubkey.key.e, pubkey.key.elen))
+  let n = base64UrlEncode(nArray)
+  let e = base64UrlEncode(eArray)
+
+  let newNonce = await self.newNonce()
+  var header = %*{"alg": Alg, "typ": "JWT", "nonce": newNonce, "url": url}
+  if needsJwk:
+    header["jwk"] = %*{"kty": "RSA", "n": n, "e": e}
+  else:
+    if self.kid.isNone:
+      raise newException(ACMEError, "no kid registered for account")
+    header["kid"] = %*(self.kid.get)
+  return header
+
+proc signedAcmeRequest(
+    self: ref ACMEAccount, url: string, payload: JsonNode, needsJwk: bool = false
+): Future[HttpClientResponseRef] {.async: (raises: [ACMEError, CancelledError]).} =
+  let acmeHeader = await self.acmeHeader(url, needsJwk)
+  var token: JWT
+  var body: JsonNode
+  try:
+    token = toJWT(%*{"header": acmeHeader, "claims": payload})
+    let derPrivKey = self.key.seckey.rsakey.getBytes.get
+    let pemPrivKey: string = pemEncode(derPrivKey, "PRIVATE KEY")
+    token.sign(pemPrivKey)
+    body = token.toFlattenedJson()
+  except Exception as exc:
+    raise newException(ACMEError, "Failed to create JWT", exc)
+  try:
+    let response = await HttpClientRequestRef
```

Comment:
nitpick: since variable `response` is not really used we can directly return here `return await ...`